jax/_src/numpy/lax_numpy.py

```diff
@@ -2098,6 +2099,24 @@ def array(object: Any, dtype: DTypeLike | None = None, copy: bool = True,
           order: str | None = "K", ndmin: int = 0) -> Array:
   if order is not None and order != "K":
     raise NotImplementedError("Only implemented for order='K'")
+  if not isinstance(object, Array):
+    # Check if object supports any of the data exchange protocols
+    # (except dlpack, see data-apis/array-api#301). If it does,
+    # consume the object as jax array and continue (but not return) so
+    # that other array() arguments get processed against the input
+    # object.
+    #
+    # Notice that data exchange protocols define dtype in the
+    # corresponding data structures and it may not be available as
+    # object.dtype. So, we'll resolve the protocols here before
+    # evaluating object.dtype.
+    if hasattr(object, '__jax_array__'):
+      object = object.__jax_array__()
+    elif xla_extension_version >= 237:
+      cai = getattr(object, "__cuda_array_interface__", None)
+      if cai is not None:
+        backend = xla_bridge.get_backend("cuda")
+        object = xc._xla.cuda_array_interface_to_buffer(cai, backend)
 
   # check if the given dtype is compatible with JAX
   dtypes.check_user_dtype_supported(dtype, "array")
```

Comment:
Let's do this check above the `__jax_array__` and `__cuda_array_interface__` checks.

In fact, I think we should put this block where the `__jax_array__` check originally appeared.